File: 0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.java
class Solution {
    public List<Integer> diffWaysToCompute(String expression) {
        List<Integer> ans=new ArrayList<>();
        for(int i=0;i<expression.length();i++){
            char c=expression.charAt(i);
            if(c=='+'||c=='-'||c=='*'){
                List<Integer> s1=diffWaysToCompute(expression.substring(0,i));
                List<Integer> s2=diffWaysToCompute(expression.substring(i+1));
                for(int j:s1){
                    for(int k:s2){
                        if(c=='+') ans.add(j+k);
                        else if(c=='-') ans.add(j-k);
                        else ans.add(j*k);
                    }
                }
            }
        }
        if(ans.isEmpty()) ans.add(Integer.parseInt(expression));
        return ans;
    }
}

File: 0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {
        vector<int> ans;
        for(int i=0;i<expression.size();i++){
            if(expression[i]=='+'||expression[i]=='-'||expression[i]=='*'){
                vector<int> s1=diffWaysToCompute(expression.substr(0,i));
                vector<int> s2=diffWaysToCompute(expression.substr(i+1));
                for(int a:s1){
                    for(int b:s2){
                        if(expression[i]=='+') ans.push_back(a+b);
                        else if(expression[i]=='-') ans.push_back(a-b);
                        else ans.push_back(a*b);
                    }
                }
            }
        }
        if(ans.empty()) ans.push_back(stoi(expression));
        return ans;
    }
};